lambda-dlq/main.tf

```diff
@@ -10,33 +13,73 @@ resource "aws_sns_topic_subscription" "lambda-dlq" {
 }
 
 resource "aws_sqs_queue" "lambda-dlq" {
-  name = "${var.name}-lambda-dlq"
+  name                      = "${var.name}-lambda-dlq"
   message_retention_seconds = 1209600 # 14d
 
   kms_master_key_id = var.kms_master_key_id
 }
 
 data "aws_iam_policy_document" "lambda-dlq" {
   statement {
-    sid = "DLQ"
-    effect = "Allow"
-    actions = ["sns:Publish"]
+    sid       = "DLQ"
+    effect    = "Allow"
+    actions   = ["sns:Publish"]
     resources = [aws_sns_topic.lambda-dlq.arn]
   }
   statement {
-    sid = "KMS"
-    effect = "Allow"
+    sid     = "KMS"
+    effect  = "Allow"
     actions = [
       "kms:GenerateDataKey",
       "kms:GenerateDataKeyPair",
       "kms:GenerateDataKeyPairWithoutPlaintext",
       "kms:GenerateDataKeyWithoutPlaintext",
       "kms:Decrypt"
     ]
-    resources = ["arn:aws:kms:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:key/${var.kms_master_key_id}"]
+    resources = [
+      "arn:aws:kms:${data.aws_region.current.name}:${data.aws_caller_identity.current.account_id}:key/${var.kms_master_key_id}"
+    ]
   }
 }
 resource "aws_iam_policy" "lambda-dlq" {
-  name = "${var.name}-lambda-dlq-policy"
+  name   = "${var.name}-lambda-dlq-policy"
   policy = data.aws_iam_policy_document.lambda-dlq.json
+}
+
+data "aws_iam_policy_document" "lambda-dlq-sns" {
+  statement {
+    effect  = "Allow"
+    actions = ["sts:AssumeRole"]
+
+    principals {
+      type        = "Service"
+      identifiers = ["sns.${data.aws_partition.current.dns_suffix}"]
+    }
+  }
+}
+
+resource "aws_iam_role" "lambda-dlq-sns" {
+  name               = "${var.name}-lambda-dlq-sns-role"
+  path               = "/"
+  assume_role_policy = data.aws_iam_policy_document.lambda-dlq-sns.json
+}
+
+data "aws_iam_policy_document" "lambda-dlq-sns-delivery-status-role-policy" {
+  statement {
+    effect  = "Allow"
+    actions = [
+      "logs:CreateLogGroup",
+      "logs:CreateLogStream",
+      "logs:PutLogEvents",
+      "logs:PutMetricFilter",
+      "logs:PutRetentionPolicy"
+    ]
+    resources = ["*"]
+  }
+}
+
+resource "aws_iam_role_policy" "lambda-dlq-sns-delivery-status-role-policy" {
```

Comment:
needs aws_iam_role_policy_attachment to attach to new role

Reply:
Here its created as an inline policy to the new IAM role using [aws_iam_role_policy](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_role_policy). Do you want me to change it to a managed policy using [aws_iam_policy](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/iam_policy) and then use aws_iam_role_policy_attachment?

Reply:
assume_role_policy is different, you need to attach this one as well

Reply:
Sorry, but isn't it already attached to the new `lambda-dlq-sns` role at line 83?

Its showing up in the console: https://us-east-1.console.aws.amazon.com/iamv2/home?region=us-east-1#/roles/details/api-private-lambda-dlq-sns-role?section=permissions

Reply:
ah, I see. yeah that's fine. I usually use aws_iam_role_policy_attachment because I'm always reusing policies. Not needed in this case.